File: backend/db/schema.py
"""Database schema using Drizzle."""
import uuid
from datetime import datetime
from drizzle_orm import asyncpg

# Users table
users = asyncpg.Table(
    "users",
    asyncpg.Column("id", asyncpg.UUID, primary_key=True, default=uuid.uuid4),
    asyncpg.Column("email", asyncpg.Varchar(255), unique=True, not_null=True),
    asyncpg.Column("password_hash", asyncpg.Varchar(255), not_null),
    asyncpg.Column("created_at", asyncpg.Timestamp, default=datetime.utcnow),
)

# Documents table
documents = asyncpg.Table(
    "documents",
    asyncpg.Column("id", asyncpg.UUID, primary_key=True, default=uuid.uuid4),
    asyncpg.Column("user_id", asyncpg.UUID, not_null),
    asyncpg.Column("type", asyncpg.Varchar(50), not_null),  # resume, job_description
    asyncpg.Column("filename", asyncpg.Varchar(255)),
    asyncpg.Column("content", asyncpg.Text),
    asyncpg.Column("file_path", asyncpg.Varchar(500)),
    asyncpg.Column("created_at", asyncpg.Timestamp, default=datetime.utcnow),
)

# Profiles table (extracted from documents)
profiles = asyncpg.Table(
    "profiles",
    asyncpg.Column("id", asyncpg.UUID, primary_key=True, default=uuid.uuid4),
    asyncpg.Column("document_id", asyncpg.UUID),
    asyncpg.Column("user_id", asyncpg.UUID, not_null),
    asyncpg.Column("profile_type", asyncpg.Varchar(50), not_null),  # candidate, job
    asyncpg.Column("name", asyncpg.Varchar(255)),
    asyncpg.Column("headline", asyncpg.Text),
    asyncpg.Column("skills", asyncpg.Text),  # JSON string
    asyncpg.Column("experience", asyncpg.Text),
    asyncpg.Column("company", asyncpg.Varchar(255)),
    asyncpg.Column("role", asyncpg.Varchar(255)),
    asyncpg.Column("requirements", asyncpg.Text),
    asyncpg.Column("confidence_score", asyncpg.Float),
    asyncpg.Column("created_at", asyncpg.Timestamp, default=datetime.utcnow),
)

# Interview contexts table
interview_contexts = asyncpg.Table(
    "interview_contexts",
    asyncpg.Column("id", asyncpg.UUID, primary_key=True, default=uuid.uuid4),
    asyncpg.Column("user_id", asyncpg.UUID, not_null),
    asyncpg.Column("resume_profile_id", asyncpg.UUID),
    asyncpg.Column("job_profile_id", asyncpg.UUID),
    asyncpg.Column("created_at", asyncpg.Timestamp, default=datetime.utcnow),
)

# Sessions table
sessions = asyncpg.Table(
    "sessions",
    asyncpg.Column("id", asyncpg.UUID, primary_key=True, default=uuid.uuid4),
    asyncpg.Column("user_id", asyncpg.UUID, not_null),
    asyncpg.Column("context_id", asyncpg.UUID),
    asyncpg.Column("agent_id", asyncpg.UUID),  # Current interviewer agent
    asyncpg.Column("status", asyncpg.Varchar(50), default="pending"),  # pending, active, paused, ended
    asyncpg.Column("started_at", asyncpg.Timestamp, default=datetime.utcnow),
    asyncpg.Column("ended_at", asyncpg.Timestamp),
    asyncpg.Column("transcript", asyncpg.JSONB, default=[]),
    asyncpg.Column("reconnect_token", asyncpg.UUID, default=uuid.uuid4),  # For session persistence
)

__all__ = ["users", "documents", "profiles", "interview_contexts", "sessions", "agent_configs", "debriefs"]

# Agent configs table (versioned interviewer personas)
agent_configs = asyncpg.Table(
    "agent_configs",
    asyncpg.Column("id", asyncpg.UUID, primary_key=True, default=uuid.uuid4),
    asyncpg.Column("name", asyncpg.Varchar(100), not_null),
    asyncpg.Column("description", asyncpg.Text),
    asyncpg.Column("prompt_template", asyncpg.Text, not_null),
    asyncpg.Column("behavior_settings", asyncpg.JSONB, default={}),
    asyncpg.Column("rubric_definition", asyncpg.JSONB, default=[]),
    asyncpg.Column("version", asyncpg.Integer, default=1),
    asyncpg.Column("is_active", asyncpg.Boolean, default=True),
    asyncpg.Column("created_at", asyncpg.Timestamp, default=datetime.utcnow),
    asyncpg.Column("updated_at", asyncpg.Timestamp, default=datetime.utcnow),
)

# Debriefs table
debriefs = asyncpg.Table(
    "debriefs",
    asyncpg.Column("id", asyncpg.UUID, primary_key=True, default=uuid.uuid4),
    asyncpg.Column("session_id", asyncpg.UUID, not_null),
    asyncpg.Column("user_id", asyncpg.UUID, not_null),
    asyncpg.Column("scores", asyncpg.JSONB, default={}),
    asyncpg.Column("feedback", asyncpg.Text),
    asyncpg.Column("evidence", asyncpg.JSONB, default=[]),
    asyncpg.Column("rubric_used", asyncpg.JSONB, default=[]),
    asyncpg.Column("generated_at", asyncpg.Timestamp, default=datetime.utcnow),
)

File: drizzle/schema.ts
import {
  boolean,
  doublePrecision,
  index,
  integer,
  jsonb,
  pgTable,
  text,
  timestamp,
  uniqueIndex,
  uuid,
  varchar,
} from "drizzle-orm/pg-core";

export const documents = pgTable("documents", {
  id: uuid("id").primaryKey(),
  userId: uuid("user_id").notNull(),
  kind: varchar("kind", { length: 50 }).notNull(),
  type: varchar("type", { length: 50 }),
  sourceType: varchar("source_type", { length: 50 }),
  filename: varchar("filename", { length: 255 }),
  mimeType: varchar("mime_type", { length: 255 }),
  storagePath: varchar("storage_path", { length: 500 }),
  filePath: varchar("file_path", { length: 500 }),
  sourceUrl: text("source_url"),
  rawText: text("raw_text"),
  content: jsonb("content").$type<Record<string, unknown>>().default({}),
  parseStatus: varchar("parse_status", { length: 50 }).default("pending"),
  createdAt: timestamp("created_at", { withTimezone: true }).defaultNow(),
});

export const profiles = pgTable("profiles", {
  id: uuid("id").primaryKey(),
  documentId: uuid("document_id").references(() => documents.id, { onDelete: "set null" }),
  userId: uuid("user_id").notNull(),
  profileType: varchar("profile_type", { length: 50 }).notNull(),
  name: varchar("name", { length: 255 }),
  headline: text("headline"),
  skills: text("skills"),
  experience: text("experience"),
  company: varchar("company", { length: 255 }),
  role: varchar("role", { length: 255 }),
  requirements: text("requirements"),
  confidenceScore: doublePrecision("confidence_score"),
  structuredJson: jsonb("structured_json").$type<Record<string, unknown>>().default({}),
  summaryText: text("summary_text"),
  createdAt: timestamp("created_at", { withTimezone: true }).defaultNow(),
});

export const agentConfigs = pgTable("agent_configs", {
  id: uuid("id").primaryKey(),
  name: varchar("name", { length: 100 }).notNull(),
  description: text("description"),
  promptTemplate: text("prompt_template").notNull(),
  behaviorSettings: jsonb("behavior_settings").$type<Record<string, unknown>>().default({}),
  rubricDefinition: jsonb("rubric_definition").$type<Record<string, unknown>[]>().default([]),
  version: integer("version").default(1),
  isActive: boolean("is_active").default(true),
  createdAt: timestamp("created_at", { withTimezone: true }).defaultNow(),
  updatedAt: timestamp("updated_at", { withTimezone: true }).defaultNow(),
}, (table) => ({
  isActiveCreatedAtIdx: index("agent_configs_is_active_created_at_idx").on(table.isActive, table.createdAt),
}));

export const interviewContexts = pgTable("interview_contexts", {
  id: uuid("id").primaryKey(),
  userId: uuid("user_id").notNull(),
  resumeProfileId: uuid("resume_profile_id").references(() => profiles.id, { onDelete: "set null" }),
  jobProfileId: uuid("job_profile_id").references(() => profiles.id, { onDelete: "set null" }),
  agentId: uuid("agent_id").references(() => agentConfigs.id, { onDelete: "set null" }),
  customInstructions: text("custom_instructions"),
  matchAnalysisJson: jsonb("match_analysis_json").$type<Record<string, unknown>>().default({}),
  createdAt: timestamp("created_at", { withTimezone: true }).defaultNow(),
}, (table) => ({
  userCreatedAtIdx: index("interview_contexts_user_created_at_idx").on(table.userId, table.createdAt),
  agentIdIdx: index("interview_contexts_agent_id_idx").on(table.agentId),
}));

export const sessions = pgTable("sessions", {
  id: uuid("id").primaryKey(),
  userId: uuid("user_id").notNull(),
  contextId: uuid("context_id").references(() => interviewContexts.id, { onDelete: "set null" }),
  agentId: uuid("agent_id").references(() => agentConfigs.id, { onDelete: "set null" }),
  status: varchar("status", { length: 50 }).default("pending"),
  startedAt: timestamp("started_at", { withTimezone: true }).defaultNow(),
  endedAt: timestamp("ended_at", { withTimezone: true }),
  transcript: jsonb("transcript").$type<Record<string, unknown>[]>().default([]),
  reconnectToken: uuid("reconnect_token").notNull(),
}, (table) => ({
  userStartedAtIdx: index("sessions_user_started_at_idx").on(table.userId, table.startedAt),
  reconnectTokenIdx: uniqueIndex("sessions_reconnect_token_idx").on(table.reconnectToken),
  contextIdIdx: index("sessions_context_id_idx").on(table.contextId),
  agentIdIdx: index("sessions_agent_id_idx").on(table.agentId),
}));

export const debriefs = pgTable("debriefs", {
  id: uuid("id").primaryKey(),
  sessionId: uuid("session_id").references(() => sessions.id, { onDelete: "cascade" }).notNull(),
  userId: uuid("user_id").notNull(),
  scores: jsonb("scores").$type<Record<string, number>>().default({}),
  feedback: text("feedback"),
  evidence: jsonb("evidence").$type<Record<string, unknown>[]>().default([]),
  rubricUsed: jsonb("rubric_used").$type<Record<string, unknown>[]>().default([]),
  generatedAt: timestamp("generated_at", { withTimezone: true }).defaultNow(),
});
